### This Project explores the NYC Taxi data in order to find a metric to measure the efficiency of the ride

## Part 1: Aggregation 

#### Question: How should we measure the effciency of the ride

The metric I propose is **Weighted Proportion of Efficient Rides** among total rides. Details on **Weight** and **Efficient Rides** are provided in the following section. 

Suppose we have an algorithm to tell us whether a ride is **efficient or not** (this algorithm will be discussed in the following steps). Also suppose that we have a **weight** for each ride (this weight will be discussed in the following steps). Using this information, we can get the overall efficiency of many rides using the following formula:

$$efficiency=\frac{\text{sum of weights for all efficient rides}}{\text{sum of weights for all rides}}$$

Now, lets discuss the algorithm to judge whether a ride is **efficient or not**. 

I believe, when it comes to the efficiency of a ride, there are the following two factors that matter: 

- **Number of people in that ride** <br>
For example, when two cars are travelling from place A to place B, if both of them are carrying only one passenger, then one of the two cars are having **inefficient ride** because we could potentially use just one car for two passengers (assuming car can carry up to 5 people).<br> 
However, if both of those cars are carrying 4 passengers, then both of the cars are having **efficient ride** because there is no way we can fit 8 passengers in one car.<br>


- **Direction of the ride (distance, time)**<br>
For example, there are two cars (car1 and car2) and both of them are carrying only one passenger. Lets assume car1 is travelling from place A to place B, car2 is travelling from place A to place C. If place B and place C are very close and it will only take very small amount of extra time for one car to drop off the passenger at place B and then drop off another passenger at place C, then we can argue that one of the cars is having **inefficient ride**. <br>
However, if place B and place C are very far and it takes a lot of extra time for one car to drop off one passenger at place B and then drop off another passenger at place C, then we can argue that both of the rides are having **efficient ride**.

Based on the above discussion, I propose the following algorithm for judging whether each ride is **efficient or not** for a set of rides that have close pick up times (pick up times are within 5 minutes interval. We have to batch process here because pick up times of the rides should be very close for us to combine them ): 

**For** each ride **do**:<br> 
&nbsp; &nbsp; &nbsp; &nbsp; **if** you can find another ride to merge (rule for merging is explained below) with this ride **then** <br> 
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; mark the current ride as **inefficient** <br> 
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; update the other ride with new number of passengers and new route <br>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; move on to the next ride <br>
&nbsp; &nbsp; &nbsp; &nbsp; **else**: <br>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; mark the current ride as **efficient** <br> 
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; move on to the next ride <br>

How do we decide if we can merge two rides? <br>

Rule to merge two rides: 
-  Car should be able to carry the passengers of two cars
- the extra time spent on the new route should be less than 10 minutes (this is a parameter of the algorithm we can change later)
- the extra distance travelled should be less than 5 miles (this is a paramater of the algorithm we can change later)

For example, lets assume there are 10 rides. To judge wether a ride is efficient or not, we compare current ride with all other 9 rides and see if it is possible to combine current ride with any of those 9 rides on the condition that the other car can fit the passengers of two cars, the **extra distance** with the new route will be less than 5 miles and **extra trip time** with the new route will be less than 10 minutes. 

Now, we have an algorithm to judge whether a ride is efficient or not. Now, lets discuss the **weight** of each ride.<br> 

It is necessary to have weight for each ride because there are some rides that are very inefficient. For example, there are two rides, both of them are inefficient based on our algorithm discussed above. However, if one of the rides has a very long trip distance, our metric should be able to reflect the fact that this ride is more inefficient than the other ride with shorter trip distance. 
Thus, I suggest using the **travel distance** as the weight for our efficiency metric. 

Now we have an algorithm to judge whether a ride is efficient or nor. We also have weight for each ride. We can use the formula provided at the beginning of our discussion to calculate the efficiency of many rides. 

#### Question: Actual Implementation

**Note**: this code is written in **Python version 2.7**; Please install all the necessary libraries to be able to successfully run the notebook. Please make sure you have **BoroughBoundaries.geojson** and **yellow_tripdata_2016-06.csv** files placed in correct directory.  

In this section, I utilize **Google maps api** and the algorithm I suggested in question 1 to label each ride as **efficient or inefficient** one batch at a time. Batch processing is necessary because two rides have to have very close pick up time in order for us to combine them. I then use the formula suggested in section 1 to calculate the overall efficiency for Manhattan. I utilize NYC borough boundaries geojson file I obtained from NYC Opendata website to map each trip to borough. 

In [1]:
import geopandas
import numpy as np
import pandas as pd
from shapely.geometry import Point
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime
from datetime import date
from datetime import timedelta
import seaborn as sns 
import googlemaps
import warnings
warnings.filterwarnings('ignore')

In [2]:
### read in the borough boundaries geojson file 
borough = geopandas.read_file("BoroughBoundaries/BoroughBoundaries.geojson")

In [3]:
borough.head(10)

boro_code      boro_name     shape_area     shape_leng  \
0         2          Bronx  1186612476.77  462958.188213   
1         5  Staten Island  1623756421.84  325960.634597   
2         3       Brooklyn  1937593021.46  738745.835869   
3         4         Queens  3045885240.47  904390.137335   
4         1      Manhattan  636602662.347  361212.479734   

                                            geometry  
0  (POLYGON ((-73.89680883223774 40.7958084451597...  
1  (POLYGON ((-74.05050806403247 40.5664220341608...  
2  (POLYGON ((-73.867061494062 40.58208797679264,...  
3  (POLYGON ((-73.83668274106707 40.5949466970158...  
4  (POLYGON ((-74.01092841268031 40.6844914725429...

In [4]:
### read in the yellow taxi trip data for 2016 June. 
df_taxi=pd.read_csv('data/yellow_tripdata_2016-06.csv')

In [5]:
### process the data, filter the data for the first full week of June 2016 (2016-06-06 to 2016-06-12)
df_taxi['tpep_pickup_datetime']=pd.to_datetime(df_taxi['tpep_pickup_datetime'])
df_taxi['tpep_dropoff_datetime']=pd.to_datetime(df_taxi['tpep_dropoff_datetime'])

In [6]:
start_date=date(year=2016,month=6,day=6)
end_date=date(year=2016,month=6,day=12)

In [7]:
df_taxi=df_taxi[(df_taxi['tpep_pickup_datetime']>start_date)&(df_taxi['tpep_dropoff_datetime']<end_date)]

##### Create Borough names from latitude and longitude using Borough Boundaries geojson files. 

In [8]:
df_taxi=df_taxi.sort_values(by='tpep_pickup_datetime')
df_taxi=df_taxi[:1000]

Here, I decided to take a sample of first 1000 trips due to the very long running time of the algorithm. Since we are making calls to the google maps api and getting trip distance and time, this process is taking very long time. We do expect that efficiency calculated from these samples might be different from overall efficiency since by taking this sample, we are effectively only considering trips within 10 minute interval. Due to limited time and resources, we decided to result of this calculation as our estimate for efficiency. In other words, the assumtion we are making when we use the sample data here is that efficiency does not change much over time for Manhattan.  

In the future, when we are not bounded by time, we do not have to take sample and calculate the efficiency for all trips. 

Note: usual method of taking random sampling does not work here because if we randomly sample trips from all trips, we are reducing the number of trips in each batch, which will give us incorrect estimation of efficiency. 

In [9]:
df_taxi.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2765159         2  2016-06-06 00:00:01   2016-06-06 00:04:05                2   
2128993         1  2016-06-06 00:00:02   2016-06-06 00:03:36                1   
2128995         2  2016-06-06 00:00:03   2016-06-06 00:00:35                1   
2765160         1  2016-06-06 00:00:03   2016-06-06 00:04:13                1   
2765161         2  2016-06-06 00:00:03   2016-06-06 00:11:18                1   

         trip_distance  pickup_longitude  pickup_latitude  RatecodeID  \
2765159           1.37        -73.980690        40.742210           1   
2128993           0.90        -73.986794        40.756310           1   
2128995           0.02        -74.004211        40.742249           1   
2765160           0.50        -73.977020        40.751965           1   
2765161           2.65        -74.006348        40.733410           1   

        store_and_fwd_flag  dropoff_longitude  dropoff_latitude  payment_type  \
2765159                  N         -73.967842         40.759029             1   
2128993                  N         -73.976067         40.763119             2   
2128995                  N         -74.001801         40.741249             2   
2765160                  N         -73.974411         40.753727             2   
2765161                  N         -74.008247         40.704369             1   

         fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
2765159          6.0    0.5      0.5        1.82           0.0   
2128993          5.0    0.5      0.5        0.00           0.0   
2128995          2.5    0.5      0.5        0.00           0.0   
2765160          4.5    0.5      0.5        0.00           0.0   
2765161         10.5    0.5      0.5        2.36           0.0   

         improvement_surcharge  total_amount  
2765159                    0.3          9.12  
2128993                    0.3          6.30  
2128995                    0.3          3.80  
2765160                    0.3          5.80  
2765161                    0.3         14.16

In [10]:
df_taxi['pickup_coordinates'] = df_taxi[['pickup_longitude', 'pickup_latitude']].values.tolist()
df_taxi['pickup_coordinates'] = df_taxi['pickup_coordinates'].apply(Point)

In [11]:
df_taxi['dropoff_coordinates'] = df_taxi[['dropoff_longitude', 'dropoff_latitude']].values.tolist()
df_taxi['dropoff_coordinates'] = df_taxi['dropoff_coordinates'].apply(Point)

In [12]:
df_taxi.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2765159         2  2016-06-06 00:00:01   2016-06-06 00:04:05                2   
2128993         1  2016-06-06 00:00:02   2016-06-06 00:03:36                1   
2128995         2  2016-06-06 00:00:03   2016-06-06 00:00:35                1   
2765160         1  2016-06-06 00:00:03   2016-06-06 00:04:13                1   
2765161         2  2016-06-06 00:00:03   2016-06-06 00:11:18                1   

         trip_distance  pickup_longitude  pickup_latitude  RatecodeID  \
2765159           1.37        -73.980690        40.742210           1   
2128993           0.90        -73.986794        40.756310           1   
2128995           0.02        -74.004211        40.742249           1   
2765160           0.50        -73.977020        40.751965           1   
2765161           2.65        -74.006348        40.733410           1   

        store_and_fwd_flag  dropoff_longitude  \
2765159                  N         -73.967842   
2128993                  N         -73.976067   
2128995                  N         -74.001801   
2765160                  N         -73.974411   
2765161                  N         -74.008247   

                             ...                       payment_type  \
2765159                      ...                                  1   
2128993                      ...                                  2   
2128995                      ...                                  2   
2765160                      ...                                  2   
2765161                      ...                                  1   

         fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
2765159          6.0    0.5      0.5        1.82           0.0   
2128993          5.0    0.5      0.5        0.00           0.0   
2128995          2.5    0.5      0.5        0.00           0.0   
2765160          4.5    0.5      0.5        0.00           0.0   
2765161         10.5    0.5      0.5        2.36           0.0   

         improvement_surcharge  total_amount  \
2765159                    0.3          9.12   
2128993                    0.3          6.30   
2128995                    0.3          3.80   
2765160                    0.3          5.80   
2765161                    0.3         14.16   

                                   pickup_coordinates  \
2765159   POINT (-73.98069000244141 40.7422103881836)   
2128993  POINT (-73.98679351806641 40.75630950927734)   
2128995  POINT (-74.00421142578125 40.74224853515625)   
2765160   POINT (-73.97702026367188 40.7519645690918)   
2765161      POINT (-74.00634765625 40.7334098815918)   

                                  dropoff_coordinates  
2765159  POINT (-73.96784210205078 40.75902938842773)  
2128993  POINT (-73.97606658935547 40.76311874389648)  
2128995  POINT (-74.00180053710938 40.74124908447266)  
2765160  POINT (-73.97441101074219 40.75372695922852)  
2765161   POINT (-74.0082473754883 40.70436859130859)  

[5 rows x 21 columns]

In [13]:
### create a dictionary to be used for transforming latitude and longitude to Borough 
boro_list=borough['boro_name'].tolist()
polygon_dict={}
for boro in boro_list:
    polygon_dict[boro]=borough[borough['boro_name']==boro]['geometry'][0] 
polygon_dict

{u'Bronx': <shapely.geometry.multipolygon.MultiPolygon at 0x10e4e9090>,
 u'Brooklyn': <shapely.geometry.multipolygon.MultiPolygon at 0x10e4e9190>,
 u'Manhattan': <shapely.geometry.multipolygon.MultiPolygon at 0x10e4e9150>,
 u'Queens': <shapely.geometry.multipolygon.MultiPolygon at 0x117174b10>,
 u'Staten Island': <shapely.geometry.multipolygon.MultiPolygon at 0x1a22ce6c50>}

In [14]:
def BoroughMapper(point):
    """
    This is a helper function used to map the latitude, longitude to Borough names 
    """
    if polygon_dict['Bronx'].contains (point):
        return 'Bronx'
    elif polygon_dict['Staten Island'].contains (point):
        return 'Staten Island'
    elif polygon_dict['Brooklyn'].contains (point):
        return 'Brooklyn'
    elif polygon_dict['Queens'].contains (point):
        return 'Queens'
    elif polygon_dict['Manhattan'].contains (point):
        return 'Manhattan'
    else:
        return 'OtherBorough'

In [15]:
### create pick up and dropoff borough. 
df_taxi['pickup_borough']=df_taxi['pickup_coordinates'].apply(BoroughMapper)
print "pick up boroughs finished ......"
df_taxi['dropoff_borough']=df_taxi['dropoff_coordinates'].apply(BoroughMapper)
print "drop off boroughs finished ......"

pick up boroughs finished ......
drop off boroughs finished ......


In [16]:
df_taxi.head(5)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2765159         2  2016-06-06 00:00:01   2016-06-06 00:04:05                2   
2128993         1  2016-06-06 00:00:02   2016-06-06 00:03:36                1   
2128995         2  2016-06-06 00:00:03   2016-06-06 00:00:35                1   
2765160         1  2016-06-06 00:00:03   2016-06-06 00:04:13                1   
2765161         2  2016-06-06 00:00:03   2016-06-06 00:11:18                1   

         trip_distance  pickup_longitude  pickup_latitude  RatecodeID  \
2765159           1.37        -73.980690        40.742210           1   
2128993           0.90        -73.986794        40.756310           1   
2128995           0.02        -74.004211        40.742249           1   
2765160           0.50        -73.977020        40.751965           1   
2765161           2.65        -74.006348        40.733410           1   

        store_and_fwd_flag  dropoff_longitude       ...        extra  mta_tax  \
2765159                  N         -73.967842       ...          0.5      0.5   
2128993                  N         -73.976067       ...          0.5      0.5   
2128995                  N         -74.001801       ...          0.5      0.5   
2765160                  N         -73.974411       ...          0.5      0.5   
2765161                  N         -74.008247       ...          0.5      0.5   

         tip_amount  tolls_amount  improvement_surcharge  total_amount  \
2765159        1.82           0.0                    0.3          9.12   
2128993        0.00           0.0                    0.3          6.30   
2128995        0.00           0.0                    0.3          3.80   
2765160        0.00           0.0                    0.3          5.80   
2765161        2.36           0.0                    0.3         14.16   

                                   pickup_coordinates  \
2765159   POINT (-73.98069000244141 40.7422103881836)   
2128993  POINT (-73.98679351806641 40.75630950927734)   
2128995  POINT (-74.00421142578125 40.74224853515625)   
2765160   POINT (-73.97702026367188 40.7519645690918)   
2765161      POINT (-74.00634765625 40.7334098815918)   

                                  dropoff_coordinates  pickup_borough  \
2765159  POINT (-73.96784210205078 40.75902938842773)       Manhattan   
2128993  POINT (-73.97606658935547 40.76311874389648)       Manhattan   
2128995  POINT (-74.00180053710938 40.74124908447266)       Manhattan   
2765160  POINT (-73.97441101074219 40.75372695922852)       Manhattan   
2765161   POINT (-74.0082473754883 40.70436859130859)       Manhattan   

        dropoff_borough  
2765159       Manhattan  
2128993       Manhattan  
2128995       Manhattan  
2765160       Manhattan  
2765161       Manhattan  

[5 rows x 23 columns]

We now mapped pick up latitude and longitude to NYC boroughs. 

In [17]:
### filter the rides to both pick up and drop off locations in Manhattan. 
df_taxi=df_taxi[(df_taxi['pickup_borough']=='Manhattan')&(df_taxi['dropoff_borough']=='Manhattan')]

In [18]:
len(df_taxi)

629

##### Utilize Google maps api to create a function that calculates time and distance for two points

In [19]:
gmaps = googlemaps.Client(key='AIzaSyD8O51IF4ORRXvWh_L2wMXYfDOY11Phdnc')
def GoogleMapTimeDistance (lat1,long1,lat2,long2):
    """
    A function that calculates time and distance for two locations. 
    """
    now = datetime.now()
    directions_result = gmaps.directions(origin=(lat1,long1),destination=(lat2,long2),mode="driving",departure_time=now)
    ### get the direction duration in seconds and device by 60 seconds to get minutes 
    duration_mins=directions_result[0]['legs'][0]['duration']['value']/60.0
    ### get the direction distance in meters and devide by 1609 to get miles 
    duration_distance=directions_result[0]['legs'][0]['distance']['value']/1609.0 
    return duration_mins, duration_distance 
    

In [20]:
### Since we can only merge two rides if their pick up time is close enough (their pick up time happened 
### within 5 mins interval), we create a 5 minute interval.  
diff_time=timedelta(minutes=5)

##### In the following section, we create batches to process the rides and label them as efficient or inefficient. 

In [21]:
### Helper function that creates a list of time intervals. This will be used to device the trips into different batches.  
def datetime_range(start, end, delta):
    current = start
    while current < end:
        yield current
        current += delta
### list of time intervals used for creating batches
dts = [pd.Timestamp(dt) for dt in 
       datetime_range(min(df_taxi['tpep_pickup_datetime']), max(df_taxi['tpep_pickup_datetime']), 
       timedelta(minutes=5))]

In [22]:
### create a list that holds batches of trips (each batch has trips that have pick up times within 5 minutes interval)
batches=[]
for i in range(len(dts)):
    cur_time=dts[i]
    cur_df=df_taxi[(df_taxi['tpep_pickup_datetime']>=cur_time)&(df_taxi['tpep_pickup_datetime']<=(cur_time+diff_time))]
    
    cur_df=cur_df.dropna()
    if len(cur_df)>0:
        batches.append(cur_df)

In [23]:
"Number of batches is", len(batches)

('Number of batches is', 2)

Currently, we have 2 batches. As explained above, each batch contains rides that have pick up times within 5 minute interval ( For example, first batch contains trips that have happened between 21:08 to 21:13) 

#### Algorithm to label trips 

The following function that takes a single batch of trips and labels each trip as **efficient** or **inefficient**. As mentioned earlier, we look at number of customers as well as extra trip time to see if a trip is inefficient. 

In [24]:
def EfficiencyGenerator(df_taxi):
    """
    Core function that generates efficiency for each trip. 
    """
    ### create list to hold efficiency label 
    efficiency=[]
    ### sort the rides by trip distance, start from the trips with shorter trip distance and see if we can find 
    ### another trip that we can merge with current trip (include current trip on the road of another trip) 
    df_taxi=df_taxi.sort_values(by='trip_distance', ascending=True)
    ### iterate through each trip and label it 
    for i in range(len(df_taxi)): 
        print "current trip is", i, "out of", len(df_taxi)
        cur_ride_efficiency='efficient'
        cur_ride=df_taxi.iloc[i]
        ### compare the current trip with all other trips 
        for j in range(i+1,len(df_taxi)):
            ### if sum of passengers for these two trips is smaller than or equal to 4 (assuming each car holds 4 passenger)
            ### see if the extra time spent is smaller than 10 minutes 
            if cur_ride['passenger_count']+df_taxi.iloc[j]['passenger_count']<=4:
                ### use google maps api to get current ride travel time and distance 
                cur_ride_mins,cur_ride_miles=GoogleMapTimeDistance (cur_ride['pickup_latitude'],cur_ride['pickup_longitude'],cur_ride['dropoff_latitude'],cur_ride['dropoff_longitude'])
                ### use google maps api to get the other ride original ride travel time and distance
                other_ride_mins,other_ride_miles=GoogleMapTimeDistance (df_taxi.iloc[j]['pickup_latitude'],df_taxi.iloc[j]['pickup_longitude'],df_taxi.iloc[j]['dropoff_latitude'],df_taxi.iloc[j]['dropoff_longitude'])
                ### get travel time and distance if two rides are combined (part1: the other ride pick up to current ride pick up) 
                combined_part1_mins,combined_part1_miles=GoogleMapTimeDistance (df_taxi.iloc[j]['pickup_latitude'],df_taxi.iloc[j]['pickup_longitude'],cur_ride['pickup_latitude'],cur_ride['pickup_longitude'])
                ### get travel time and distance if two rides are combined (part2: current ride drop off to the other ride drop off )
                combined_part2_mins,combined_part2_miles=GoogleMapTimeDistance (cur_ride['dropoff_latitude'],cur_ride['dropoff_longitude'], df_taxi.iloc[j]['dropoff_latitude'],df_taxi.iloc[j]['dropoff_longitude'])
                ### if extra time spent is less than 10 minutes compared to the maximum of original two rides, we label current trip as inefficient
                if (combined_part1_mins+cur_ride_mins+combined_part2_mins)-max(cur_ride_mins,other_ride_mins)<=10:
                    ### label current trip as inefficient, break the inner loop 
                    cur_ride_efficiency='inefficient'
                    break
        ### record efficiency of current ride 
        efficiency.append(cur_ride_efficiency)
    ### return efficiency for all rides within current batch
    return efficiency        

In [25]:
### process each batch at a time and create final dataframe that holds efficiency information. 
result_df=pd.DataFrame()
for i,cur_df in enumerate(batches):
    print "********************************"
    print "Current batch is", i, "out of", len(batches) 
    print "********************************"
    cur_efficiency=EfficiencyGenerator(cur_df)
    cur_df['efficiency']=cur_efficiency
    result_df=pd.concat([result_df,cur_df]) 

********************************
Current batch is 0 out of 2
********************************
current trip is 0 out of 432
current trip is 1 out of 432
current trip is 2 out of 432
current trip is 3 out of 432
current trip is 4 out of 432
current trip is 5 out of 432
current trip is 6 out of 432
current trip is 7 out of 432
current trip is 8 out of 432
current trip is 9 out of 432
current trip is 10 out of 432
current trip is 11 out of 432
current trip is 12 out of 432
current trip is 13 out of 432
current trip is 14 out of 432
current trip is 15 out of 432
current trip is 16 out of 432
current trip is 17 out of 432
current trip is 18 out of 432
current trip is 19 out of 432
current trip is 20 out of 432
current trip is 21 out of 432
current trip is 22 out of 432
current trip is 23 out of 432
current trip is 24 out of 432
current trip is 25 out of 432
current trip is 26 out of 432
current trip is 27 out of 432
current trip is 28 out of 432
current trip is 29 out of 432
current trip is 

current trip is 265 out of 432
current trip is 266 out of 432
current trip is 267 out of 432
current trip is 268 out of 432
current trip is 269 out of 432
current trip is 270 out of 432
current trip is 271 out of 432
current trip is 272 out of 432
current trip is 273 out of 432
current trip is 274 out of 432
current trip is 275 out of 432
current trip is 276 out of 432
current trip is 277 out of 432
current trip is 278 out of 432
current trip is 279 out of 432
current trip is 280 out of 432
current trip is 281 out of 432
current trip is 282 out of 432
current trip is 283 out of 432
current trip is 284 out of 432
current trip is 285 out of 432
current trip is 286 out of 432
current trip is 287 out of 432
current trip is 288 out of 432
current trip is 289 out of 432
current trip is 290 out of 432
current trip is 291 out of 432
current trip is 292 out of 432
current trip is 293 out of 432
current trip is 294 out of 432
current trip is 295 out of 432
current trip is 296 out of 432
current 

current trip is 98 out of 197
current trip is 99 out of 197
current trip is 100 out of 197
current trip is 101 out of 197
current trip is 102 out of 197
current trip is 103 out of 197
current trip is 104 out of 197
current trip is 105 out of 197
current trip is 106 out of 197
current trip is 107 out of 197
current trip is 108 out of 197
current trip is 109 out of 197
current trip is 110 out of 197
current trip is 111 out of 197
current trip is 112 out of 197
current trip is 113 out of 197
current trip is 114 out of 197
current trip is 115 out of 197
current trip is 116 out of 197
current trip is 117 out of 197
current trip is 118 out of 197
current trip is 119 out of 197
current trip is 120 out of 197
current trip is 121 out of 197
current trip is 122 out of 197
current trip is 123 out of 197
current trip is 124 out of 197
current trip is 125 out of 197
current trip is 126 out of 197
current trip is 127 out of 197
current trip is 128 out of 197
current trip is 129 out of 197
current tr

In [32]:
result_df[['VendorID','tpep_pickup_datetime','tpep_dropoff_datetime','passenger_count','trip_distance','pickup_borough','dropoff_borough','efficiency']].head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2765159         2  2016-06-06 00:00:01   2016-06-06 00:04:05                2   
2128993         1  2016-06-06 00:00:02   2016-06-06 00:03:36                1   
2128995         2  2016-06-06 00:00:03   2016-06-06 00:00:35                1   
2765160         1  2016-06-06 00:00:03   2016-06-06 00:04:13                1   
2765161         2  2016-06-06 00:00:03   2016-06-06 00:11:18                1   

         trip_distance pickup_borough dropoff_borough   efficiency  
2765159           1.37      Manhattan       Manhattan  inefficient  
2128993           0.90      Manhattan       Manhattan  inefficient  
2128995           0.02      Manhattan       Manhattan  inefficient  
2765160           0.50      Manhattan       Manhattan  inefficient  
2765161           2.65      Manhattan       Manhattan  inefficient

As we can see, we have labeled each ride as efficient or inefficient. Now, lets look at the efficiency rate: 

In [28]:
result_df['efficiency'].describe()

count             629
unique              2
top       inefficient
freq              471
Name: efficiency, dtype: object

In [29]:
efficient_df=result_df[result_df['efficiency']=='efficient']
inefficient_df=result_df[result_df['efficiency']=='inefficient']

In [35]:
### effficiency without considering the weight
float(len(efficient_df))/(len(inefficient_df)+len(efficient_df))

0.25119236883942764

Without considering the weight, efficiency rate for these trips is 25.12%. 

In [36]:
### considering the weight, the efficiency for these trips 
a=efficient_df['trip_distance'].sum()
b=inefficient_df['trip_distance'].sum()

In [37]:
a/(a+b)

0.2547892198272338

With considering the trip distance as weight, the efficiency rate for these trips is 25.48%. 

As we can see from the calculation above, the overall efficiency for Manhattan is 25.48% <br> 
In terms of the complexity of the algorithm, we have used two for loops in the implementation, thus the complexity is O(N^2) (O n square). 

#### Question 3 

As we can see from our analysis in question 2, the overall efficiency for manhattan is around 25%. This is a very low number. This means there is huge opportunity for us to provide ride share service and improve that efficiency through our service.

### Future work for Part 1 

We believe the result of our calculation can be further improved with the following future work:

- Instead of sampling the data, calculate the result on the entire dataset. For this, we just need to have better computing resources and let the algorithm run longer time. This way, we will have a better estimation of efficiency.


- Improve the core labeling algorithm (function EfficiencyGenerator here).As of now, both of our implementation and algorithm itself is simple. One such potential improvement is that actually update the rides (number of passengers, trip route) as we process rides in batches. In our implementation of the algorithm, we did not implement this updating process. 


- Visualization of efficiency over time as well as across different locations. Again, since we are using google maps api and this was a bottle neck in our running time of the algorithm, we did not have enough time to generate efficiency rate for different time horizon (hourly eifficienct rate) or across different locations (across different zipcode areas). As the next step, when we are not bounded by time, we could generate those results and visualize the result. 

## Part 2: Statistics

#### Question 1 (part a)

I suggest we calculate **Weighted MAPE** (mean absolute percentage error) and **Weighted Bias** to compare the performance of this forecasting model. We weight the MAPE and Bias by number of rides in each hour to get the weighted MAPE and weighted Bias for a 24 hour period. 

Since the algorithm is predicting on hourly basis, we have to penalize the algorithm more if the algorithm predicts wrong during hours with a lot of rides (this is the reason we weight the MAPE and Bias by number of rides in each hour).

I believe these metrics (weighted MAPE and weighted Bias) are comparable across cities of varying size. Some cities might heave a lot of rides and there might be huge fluctuation in the number of rides across different time. However, if our machine learning model is trained such that features related to city size (city population, city average income, city population average age, number of cars , ect) are included in the model, we expect our model to have similar performance across different cities with different size. 

#### Question 1 (part b)

It is hard to conclude that one algorithm is better than the other by simply looking at the difference between the performance of the algorithm without huge number of tests. We suggest running the current and recommended algorithm on exact same dataset, and conduct a staitstical comparison test on the performance result. The detailed steps are as below: 

- Run current algorithm on n sets of datasets and generate n sets of weighted MAPE and weighted Bias (from n sets of train/test datasets, generate n sets of weighted MAPE and weighted Bias). 
- Run recommended algorithm on same n sets of datasets and generate n sets of weighted MAPE and weighted Bias.  
- Compare the mean of weighted MAPE using Welch's t-test. Calculate p value and see if there is significant difference between mean weighted MAPEs of two groups. 

If the result of above Welch's t-test shows that our recommended algorithm is better than current algorithm (weighted MAPE is lower for recommended algorithm and p value shows the result is statistically significant), we recommend adoping the suggested algorithm. 

#### Question 2 

From the statement of the problem I am assuming 0.224 to 0.439 is the 95% confidence interval for the proportion of excellent music. Based on the formula from confidence Interval for Population Propotion: <br>
                    $$ p \pm Zα/2 * \sqrt{p * (1-p)\div n} $$ <br>

where p is the sample proportion, n is the sample size and Zα/2 is the critical value, <br> 

We want to calculate n in this formula. From he Z-table, we can find Z = 1.96. p here is calculated for population propotion who said the rides have excellent music, which is:
                    $$ p = (0.439 + 0.224)\div 2 = 0.3315 $$ <br>
Now, we could calculate for n:<br>
$$ 0.3315 + 1.96 * \sqrt{ 0.3315 * (1- 0.3315) \div n} = 0.439 $$ <br>
$$ n \approx 73.668 = 74 $$ <br>
So the orginal sample size is 74 based on our assumptions.   